ksd: Support NameServerIP

oshoval opened the pull request:
Add support of NameServerIP field to HCO CR spec (`kubeSecondaryDNSNameServerIP`).
In case the value is set, and KSD is deployed
it will be populated to CNAO CR and from there to KSD deployment.

**Reviewer Checklist**
<!-- Check [Expectations from a PR](/CONTRIBUTING.md#expectations-from-a-pr) for the details -->

> Reviewers are supposed to review the PR for every aspect below one by one. To check an item means the PR is either "OK" or "Not Applicable" in terms of that item. All items are supposed to be checked before merging a PR. 

- [ ] PR Message
- [ ] Commit Messages
- [ ] How to test
- [ ] Unit Tests
- [ ] Functional Tests
- [ ] User Documentation
- [ ] Developer Documentation
- [ ] Upgrade Scenario
- [ ] Uninstallation Scenario
- [ ] Backward Compatibility
- [ ] Troubleshooting Friendly
  
**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
None
```

orenc1 commented on docs/cluster-configuration.md:
```diff
@@ -166,6 +166,9 @@ For additional information, see here: [KubeSecondaryDNS](https://github.com/kube
 **Default**: `false`
 
+In order to set KSD's NameServerIP, set it on HyperConverged CR spec.KubeSecondaryDNSNameServerIP field.
+Default: empty string. Value is a string representation of IPv4.
```
i think the value for that field should be validated at it is a valid IPv4 address, and reject update/creation of HCO CR otherwise.
the user should get immediate feedback if the value is invalid.

oshoval replied:
Done

orenc1 commented on controllers/operands/networkAddons.go:
```diff
@@ -149,6 +150,12 @@ func NewNetworkAddons(hc *hcov1beta1.HyperConverged, opts ...string) (*networkad
 		}
+		if nameServerIP := *hc.Spec.KubeSecondaryDNSNameServerIP; nameServerIP != "" {
+			if !net.IsIPv4String(nameServerIP) {
+				return nil, errors.New("kubeSecondaryDNSNameServerIP isn't a valid IPv4")
```
are you sure the error will be shown to the user in realtime, or just CNAO CR won't be created and this will be shown only in the HCO logs?

oshoval replied:
The ValidateCreate would fail

nunnatsa commented on controllers/operands/networkAddons.go:
```diff
@@ -149,6 +150,12 @@ func NewNetworkAddons(hc *hcov1beta1.HyperConverged, opts ...string) (*networkad
 		if baseDomain != "" {
 			cnaoSpec.KubeSecondaryDNS.Domain = baseDomain
 		}
+		if nameServerIP := *hc.Spec.KubeSecondaryDNSNameServerIP; nameServerIP != "" {
```
Is it working when the field is nil? I think it will cause segmentation violation error. See here for example: https://go.dev/play/p/UjC-oXlIn-e

I think we must check for nil before accessing this field.

nunnatsa commented on controllers/operands/networkAddons_test.go:
```diff
@@ -540,9 +540,11 @@ var _ = Describe("CNA Operand", func() {
 				existingCNAO.Spec.KubeSecondaryDNS = &networkaddonsshared.KubeSecondaryDNS{}
 			}
 
+			kubeSecondaryDNSNameServerIP := "127.0.0.1"
 			if o.setFeatureGate {
 				deployKubeSecondaryDNS := o.featureGateValue
 				hco.Spec.FeatureGates.DeployKubeSecondaryDNS = &deployKubeSecondaryDNS
+				hco.Spec.KubeSecondaryDNSNameServerIP = &kubeSecondaryDNSNameServerIP
```
nit: you can use `pointer.String()` for that, e.g.
```suggestion
			if o.setFeatureGate {
				deployKubeSecondaryDNS := o.featureGateValue
				hco.Spec.FeatureGates.DeployKubeSecondaryDNS = &deployKubeSecondaryDNS
				hco.Spec.KubeSecondaryDNSNameServerIP = pointer.String("127.0.0.1")
```

oshoval replied:
but i also use the value on the EXPECT

nunnatsa commented on docs/cluster-configuration.md:
```diff
@@ -166,6 +166,9 @@ For additional information, see here: [KubeSecondaryDNS](https://github.com/kube
 
 **Default**: `false`
 
+In order to set KSD's NameServerIP, set it on HyperConverged CR spec.KubeSecondaryDNSNameServerIP field.
+Default: empty string. Value is a string representation of IPv4.
```
Please add an example, as most of the sections in this document have.

nunnatsa reviewed (changes requested):
Added a few comments. One of them seems like a real bug. please check.

oshoval commented:
Addressed comments
and extracted logic to a function, which validates the value even if KSD is not enabled
so if the changes of the CR are incremental it will still validate it on each step

nunnatsa commented on docs/cluster-configuration.md:
```diff
@@ -166,6 +166,9 @@ For additional information, see here: [KubeSecondaryDNS](https://github.com/kube
 
 **Default**: `false`
 
+In order to set KSD's NameServerIP, set it on HyperConverged CR spec.KubeSecondaryDNSNameServerIP field.
+Default: empty string. Value is a string representation of IPv4 (i.e "127.0.0.1").
+
```
We have section for each field. I understand that the `deployKubeSecondaryDNS` FG and the new `NameServerIP` are connected, and that there is no point to write the same documentation twice, but we could add a new section for the `NameServerIP` with a reference to this section.

Also, still missing an example (and there is also no example to the `deployKubeSecondaryDNS` FG).

nunnatsa commented on docs/cluster-configuration.md:
```diff
@@ -166,6 +166,9 @@ For additional information, see here: [KubeSecondaryDNS](https://github.com/kube
 
 **Default**: `false`
 
+In order to set KSD's NameServerIP, set it on HyperConverged CR spec.KubeSecondaryDNSNameServerIP field.
```
`spec.KubeSecondaryDNSNameServerIP` ==> `spec.kubeSecondaryDNSNameServerIP` (lowercase "k")

oshoval replied:
yes saw thanks

oshoval commented:
addressed comments
moved to its own section in readme

nunnatsa commented on docs/cluster-configuration.md:
```diff
@@ -636,6 +636,22 @@ spec:
+## KubeSecondaryDNS Name Server IP
+In order to set KSD's NameServerIP, set it on HyperConverged CR under spec.kubeSecondaryDNSNameServerIP field.
+Default: empty string. Value is a string representation of IPv4 (i.e "127.0.0.1").
+For more info see "deployKubeSecondaryDNS Feature Gate".
```
Could you please make it a link?:
```
For more info see [deployKubeSecondaryDNS Feature Gate](#deploykubesecondarydns-feature-gate).
```

oshoval commented:
addressed comments

nunnatsa commented:
/test hco-e2e-upgrade-index-sno-aws

nunnatsa commented:
/approve
/override coverage/coveralls
